### PR TITLE
FFI : Add a method in the client to log in using JWT

### PR DESCRIPTION
**Pull request by aristopratt**
Expose a kotlin method for logging in with JWT. 
example usage : This will help users to login to synapse using third party libraries

Tested on the local element app

- [ ] Public API changes documented in changelogs (optional)

Signed-off-by: 
Pratik Deshpande <prattspective@gmail.com>

### CODE REVIEW

File: bindings/matrix-sdk-ffi/src/client.rs [hunk 1]
```diff
@@ -295,6 +295,32 @@ impl Client {
...
+    pub async fn login_with_jwt(
+        &self,
+        jwt: String,
+        initial_device_name: Option<String>,
+        device_id: Option<String>,
+    ) -> Result<(), ClientError> {
+
+        let data: Map<String, Value> = [("token".to_owned(), jwt.into())]
+            .into_iter()
+            .collect();
```
Comment (stefanceriu) [inline bindings/matrix-sdk-ffi/src/client.rs:308]: You could also write this as `let data = json!({ "token": jwt }).as_object().unwrap().clone();` which might be a bit cleaner.

**Review by stefanceriu — CHANGES_REQUESTED** [review]
Thanks for this. It looks good to me but I left a few comments and [formatting](https://github.com/matrix-org/matrix-rust-sdk/actions/runs/11119339130/job/30894353847?pr=4056) needs a bit of work

File: bindings/matrix-sdk-ffi/src/client.rs (hunk 1 above)
Comment (aristopratt) [reply bindings/matrix-sdk-ffi/src/client.rs:308]: sure will make these changes

File: bindings/matrix-sdk-ffi/src/client.rs [hunk 2]
```diff
@@ -295,6 +295,32 @@ impl Client {
         Ok(())
     }
 
+    /// Login with JWT
+    pub async fn login_with_jwt(
```
[the remark this answers is not in the document]
Comment (aristopratt) [reply bindings/matrix-sdk-ffi/src/client.rs:299]: Sure, you're suggesting to add comments and links to this following method?

[the remark this answers is not in the document]
Comment (stefanceriu) [reply bindings/matrix-sdk-ffi/src/client.rs]: Yes please, have a look at the underlying `login_custom` as an example.

**Review by aristopratt — COMMENTED** [review]
Can someone please help me with why the fixup comment job is failing?
Also even after running cargo clippy, the formatting is failing. 
Need some help to figure it out. I would appreciate it.

**Comment by poljar** [comment]
> Can someone please help me with why the fixup comment job is failing? 

That's normal, it's just a reminder to squash the fixup commits before we merge.

> Also even after running cargo clippy, the formatting is failing. Need some help to figure it out. I would appreciate it.

You can run `cargo xtask ci fixup fmt`, we use Rust nightly for formatting.

**Comment by aristopratt** [comment]
> > Can someone please help me with why the fixup comment job is failing?
> 
> That's normal, it's just a reminder to squash the fixup commits before we merge.
> 
> > Also even after running cargo clippy, the formatting is failing. Need some help to figure it out. I would appreciate it.
> 
> You can run `cargo xtask ci fixup fmt`, we use Rust nightly for formatting.



> we use Rust nightly for formatting.

the fixup fmt command is failing. :(

**Review by aristopratt — COMMENTED** [review]
Apologies for the long commit trail. 

Let me know if i should squash all into one/ or should we squash and merge from here when required.

File: bindings/matrix-sdk-ffi/src/client.rs [hunk 3]
```diff
@@ -304,19 +304,14 @@ impl Client {
         initial_device_name: Option<String>,
         device_id: Option<String>,
     ) -> Result<(), ClientError> {
-
```
Comment (stefanceriu) [inline bindings/matrix-sdk-ffi/src/client.rs]: Are you sure it wasn't only complaining about this first one? The others were actually useful for readability.

Comment (aristopratt) [reply bindings/matrix-sdk-ffi/src/client.rs]: Youre right, it was only complaining about the first space. Hm, i redid with the readability and submitted it. 
Thanks for the help!

Comment (aristopratt) [reply bindings/matrix-sdk-ffi/src/client.rs]: Can you verify now? I have made those changes

Comment (stefanceriu) [reply bindings/matrix-sdk-ffi/src/client.rs]: Yep yep, looks good to me 👍

**Comment by stefanceriu** [comment]
Okay, looks good to me 👍  Please go ahead and squash everything into 1 commit and let's get this merged!

**Review by stefanceriu — APPROVED** [review]

**Comment by stefanceriu** [comment]
Nicely done, thank you for the contribution!

**Comment by aristopratt** [comment]
> Nicely done, thank you for the contribution!

Thanks for your guidance! Would love to contribute again in the future

**Pull request merged**